bigtable_input.py

```diff
@@ -239,28 +269,38 @@ def parse(r):
         return sorted([parse(r) for r in rows], key=operator.itemgetter(0))
 
     def delete_row_range(self, format_str, start_game, end_game):
+        global _pool
+        if _pool is None:
+            _pool = Pool(16)
         row_keys = make_single_array(
             self.tf_table.keys_by_range_dataset(
                 format_str.format(start_game),
                 format_str.format(end_game)))
         row_keys = list(row_keys)
+        if not row_keys:
+            utils.dbg('No rows left for games %d..%d' % (
+                start_game, end_game))
+            return
         utils.dbg('Deleting %d rows:  %s..%s' % (
             len(row_keys), row_keys[0], row_keys[-1]))
-        rows = [self.bt_table.row(k) for k in row_keys]
+        batches = []
         batch = []
-        def mutate_batch():
-            nonlocal batch
-            utils.dbg('...deleting batch %s..%s' % (
-                batch[0].row_key, batch[-1].row_key))
-            self.bt_table.mutate_rows(batch)
+        total_keys = len(row_keys)
+        table_name = self.bt_table.name
+        while row_keys:
+            # Reversing while batching, to trim from end
+            batch.append(row_keys.pop())
+            if len(batch) >= (bigtable.row.MAX_MUTATIONS):
+                batches.append((table_name, batch))
+                batch = []
+        if batch:
+            batches.append((table_name, batch))
             batch = []
-        while rows:
-            r = rows.pop()
-            r.delete()
-            batch.append(r)
-            if len(batch) >= bigtable.row.MAX_MUTATIONS:
-                mutate_batch()
-        mutate_batch()
+        assert sum([len(b) for t, b in batches]) == total_keys
+        utils.dbg('Deleting %d batches, total of %d keys' % (
+            len(batches), total_keys))
+        # The below is only possible with threaded pool.
+        time_taken = _pool.map(delete_rows, batches)
```

Comment:
I don't think time_taken is measuring anything.

andrew has something like
`with dbg_util_something.logged_timer("blah"):`
that helps with measuring time

Reply:
Removed the per-process timings